pkg/backend/snapshot.go

```diff
@@ -168,21 +168,79 @@ type sameSnapshotMutation struct {
 	manager *SnapshotManager
 }
 
+// mustWrite returns true if any semantically meaningful difference exists between the old and new states of a same
+// step that forces us to write the checkpoint. If no such difference exists, the checkpoint write that corresponds to
+// this step can be elided.
+func (ssm *sameSnapshotMutation) mustWrite(step deploy.Step) bool {
```

Comment:
you could pass `old` and `new` instead of the whole step, since we're asserting it's going to be an `OpSame` anyway.

Reply:
I'll move this assertion into the caller.